src/librustc_const_eval/_match.rs

```diff
@@ -628,10 +642,15 @@ pub fn is_useful<'p, 'a: 'p, 'tcx: 'a>(cx: &mut MatchCheckCtxt<'a, 'tcx>,
         // feature flag is not present, so this is only
         // needed for that case.
 
-        let is_privately_empty =
+        let mut is_privately_empty =
             all_ctors.is_empty() && !cx.is_uninhabited(pcx.ty);
         debug!("missing_ctors={:?} is_privately_empty={:?}", missing_ctors,
                is_privately_empty);
+
+        if cx.is_non_exhaustive(pcx.ty) && !cx.is_local(pcx.ty) {
```

Comment:
It's confusing to mutate this variable, especially *after* the debug println

Maybe have
```Rust
let is_privately_empty =
    all_ctors.is_empty() && !cx.is_uninhabited(pcx.ty);
let is_declared_nonexhaustive =
    cx.is_non_exhaustive(pcx.ty) && !cx.is_local(pcx.ty);

// For privately empty and nonexhaustive enums, we work as
// if there were an "extra" `_` constructor for the type, so we
// can never match over all constructors.
let is_nonexhaustive =
    is_privately_empty || is_declared_nonexhaustive;
if missing_ctors.is_empty() && !is_nonexhaustive
```

Reply:
Fixed this, thanks.